File: src/main/java/com/kingmang/lazurite/parser/pars/VisitorUtils.java
package com.kingmang.lazurite.parser.pars;

import com.kingmang.lazurite.runtime.LZR.LZRNumber;
import com.kingmang.lazurite.base.Types;
import com.kingmang.lazurite.runtime.Value;
import com.kingmang.lazurite.parser.ast.BinaryExpression;
import com.kingmang.lazurite.parser.ast.ConditionalExpression;
import com.kingmang.lazurite.parser.ast.IncludeStatement;
import com.kingmang.lazurite.parser.ast.Node;
import com.kingmang.lazurite.parser.ast.Statement;
import com.kingmang.lazurite.parser.ast.UnaryExpression;
import com.kingmang.lazurite.parser.ast.ValueExpression;
import com.kingmang.lazurite.parser.ast.VariableExpression;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public final class VisitorUtils {

    private VisitorUtils() { }

    public static boolean isValue(Node node) {
        return (node instanceof ValueExpression);
    }

    public static boolean isVariable(Node node) {
        return (node instanceof VariableExpression);
    }

    public static Statement includeProgram(IncludeStatement s) {
        if (!isValue(s)) return null;
        try {
            return s.loadProgram(s.expression.eval().asString());
        } catch (IOException ex) {
            return null;
        }
    }

    public static boolean isIntegerValue(Node node, int valueToCheck) {
        if (!isValue(node)) return false;

        final Value value = ((ValueExpression) node).value;
        if (value.type() != Types.NUMBER) return false;

        final Number number = ((LZRNumber) value).raw();
        if ( (number instanceof Integer) || (number instanceof Short) || (number instanceof Byte)) {
            return number.intValue() == valueToCheck;
        }
        return false;
    }

    public static boolean isValueAsInt(Node node, int valueToCheck) {
        if (!isValue(node)) return false;

        final Value value = ((ValueExpression) node).value;
        if (value.type() != Types.NUMBER) return false;

        return value.asInt() == valueToCheck;
    }

    public static boolean isConstantValue(Node node) {
        if (!isValue(node)) return false;

        final int type = ((ValueExpression) node).value.type();
        return ( (type == Types.NUMBER) || (type == Types.STRING) );
    }

    public static boolean isSameVariables(Node n1, Node n2) {
        if (isVariable(n1) && isVariable(n2)) {
            final VariableExpression v1 = (VariableExpression) n1;
            final VariableExpression v2 = (VariableExpression) n2;
            return v1.name.equals(v2.name);
        }
        return false;
    }

    public static Set<String> operators() {
        final Set<String> operators = new HashSet<>();
        for (BinaryExpression.Operator op : BinaryExpression.Operator.values()) {
            operators.add(op.toString());
        }
        for (UnaryExpression.Operator op : UnaryExpression.Operator.values()) {
            operators.add(op.toString());
        }
        for (ConditionalExpression.Operator op : ConditionalExpression.Operator.values()) {
            operators.add(op.getName());
        }
        return operators;
    }
}

File: src/main/java/com/kingmang/lazurite/patterns/visitor/VisitorUtils.kt
package com.kingmang.lazurite.patterns.visitor

import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.parser.ast.Node
import com.kingmang.lazurite.parser.ast.expressions.BinaryExpression
import com.kingmang.lazurite.parser.ast.expressions.ConditionalExpression
import com.kingmang.lazurite.parser.ast.expressions.UnaryExpression
import com.kingmang.lazurite.parser.ast.expressions.ValueExpression
import com.kingmang.lazurite.parser.ast.expressions.VariableExpression
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.parser.ast.statements.UsingStatement
import com.kingmang.lazurite.runtime.values.LzrNumber
import java.io.IOException


object VisitorUtils {
    @JvmStatic
    fun isValue(node: Node?): Boolean {
        return (node is ValueExpression)
    }

    @JvmStatic
    fun isVariable(node: Node?): Boolean {
        return (node is VariableExpression)
    }
    @JvmStatic
    fun loadLzrProgram(s: UsingStatement): Statement? {
        if (!isValue(s)) return null
        return try {
            s.loadLzrLibrary(s.expression.eval().asString())
        } catch (ex: IOException) {
            null
        }
    }
    @JvmStatic
    fun isIntegerValue(node: Node, valueToCheck: Int): Boolean {
        if (!isValue(node)) return false

        val value = (node as ValueExpression).value
        if (value.type() != Types.NUMBER) return false

        val number = (value as LzrNumber).raw()
        if ((number is Int) || (number is Short) || (number is Byte)) {
            return number.toInt() == valueToCheck
        }
        return false
    }
    @JvmStatic
    fun isValueAsInt(node: Node, valueToCheck: Int): Boolean {
        if (!isValue(node)) return false

        val value = (node as ValueExpression).value
        if (value.type() != Types.NUMBER) return false

        return value.asInt() == valueToCheck
    }
    @JvmStatic
    fun isConstantValue(node: Node): Boolean {
        if (!isValue(node)) return false

        val type = (node as ValueExpression).value.type()
        return ((type == Types.NUMBER) || (type == Types.STRING))
    }
    @JvmStatic
    fun isSameVariables(n1: Node, n2: Node): Boolean {
        if (isVariable(n1) && isVariable(n2)) {
            val v1 = n1 as VariableExpression
            val v2 = n2 as VariableExpression
            return v1.name == v2.name
        }
        return false
    }
    @JvmStatic
    fun operators(): Set<String> {
        val operators: MutableSet<String> = HashSet()
        for (op in BinaryExpression.Operator.entries.toTypedArray()) {
            operators.add(op.toString())
        }
        for (op in UnaryExpression.Operator.entries.toTypedArray()) {
            operators.add(op.toString())
        }
        for (op in ConditionalExpression.Operator.entries.toTypedArray()) {
            operators.add(op.name)
        }
        return operators
    }
}
